tools/dump_cinit.lua

```diff
@@ -552,18 +522,62 @@ local function dump_defs_c(symtab, basename, namespace, append)
   append("};\n")
   append("#endif\n\n")
 
+  append("static const upb_msgdef *refm(const upb_msgdef *m, const void *owner) {\n")
+  append("  upb_msgdef_ref(m, owner);\n")
+  append("  return m;\n")
+  append("}\n\n")
+  append("static const upb_enumdef *refe(const upb_enumdef *e, const void *owner) {\n")
+  append("  upb_enumdef_ref(e, owner);\n")
+  append("  return e;\n")
+  append("}\n\n")
+
+  append("/* Public API. */\n")
+
+  for m in linktab:objs(upb.DEF_MSG) do
+    append("const upb_msgdef *upbdefs_%s_get(const void *owner)" ..
+           " { return refm(%s, owner); }\n",
+           to_cident(m:full_name()), linktab:addr(m))
+  end
+
+  append("\n")
+
+  for e in linktab:objs(upb.DEF_ENUM) do
+    append("const upb_enumdef *upbdefs_%s_get(const void *owner)" ..
+           " { return refe(%s, owner); }\n",
+           to_cident(e:full_name()), linktab:addr(e))
+  end
+
   return linktab
 end
 
-local function dump_defs_for_type(format, defs, namespace, append)
+local function dump_defs_for_type(format, defs, append)
   local sorted = sorted_defs(defs)
   for _, def in ipairs(sorted) do
-    append(format, namespace, to_cident(def:full_name()), def:full_name())
+    append(format, to_cident(def:full_name()), def:full_name())
   end
 
   append("\n")
 end
 
+local function make_children_map(file)
+  -- Maps full_name -> children.
```

Comment:
Both are true, key is a prefix (foo.bar.Baz) which can be either a package or a message with nested defs. Expanded slightly to reflect this.